ykcompile/src/lib.rs

```diff
@@ -633,6 +633,86 @@ impl<TT> TraceCompiler<TT> {
         Location::Register(dst)
     }
 
+    fn c_cast(&mut self, op: &Operand, cty: &Ty) -> Location {
+        match op {
+            Operand::Place(p) => {
+                let (loc, ty) = self.place_to_location(p, false);
+                let temp = self.create_temporary();
+                match loc {
+                    Location::Register(reg) => {
+                        match cty.size() {
+                            1 => {
+                                dynasm!(self.asm
+                                    ; mov Rb(temp), Rb(reg)
+                                );
+                            }
+                            2 => match ty.size() {
+                                1 => {
+                                    dynasm!(self.asm
+                                        ; movzx Rw(temp), Rb(reg)
+                                    );
+                                }
+                                2 | 4 | 8 => {
+                                    dynasm!(self.asm
+                                        ; mov Rw(temp), Rw(reg)
```

Comment:
You should prefer 32bit operations over 8 and 16bit operations to prevent partial register stalls.

Reply:
How would I then make sure that the operand is cast down?

Reply:
Got confused. I though this was casting to a bigger size. If it is to a smaller size, you can make it a nop and just say that the upper bits of a register are undefined. This is the model Cranelift chose.

Reply:
Why a nop? I still need to move the value though, don't I, e.g. `let a = b as u8` where `b` is `u16` should result in `mov Rb(a), Rb(b)`.

Reply:
I think I get it now. The cast does nothing, and then the `store` function handles the move.

Reply:
Actually, we can't currently do a `nop` here, since the `store` function doesn't care about the size and always moves quadwords. We probably want to change the `store` function. But since everything is going to change with the new IR, we keep it as is for now.